File: android/app/src/main/java/com/vidacotidiana/app/core/network/SharingApi.kt
package com.vidacotidiana.app.core.network

import kotlinx.serialization.Serializable
import retrofit2.http.Body
import retrofit2.http.DELETE
import retrofit2.http.GET
import retrofit2.http.POST
import retrofit2.http.Path

@Serializable
data class ReminderShare(
    val id: String,
    val reminderId: String,
    val collaboratorUserId: String,
    val status: String,
    val createdAt: String,
    val revokedAt: String? = null,
)

@Serializable
data class Invitation(
    val id: String,
    val reminderId: String,
    val invitedEmail: String? = null,
    val status: String,
    val expiresAt: String,
    val createdAt: String,
)

@Serializable
data class SharesAndInvitationsResponse(
    val shares: List<ReminderShare>,
    val invitations: List<Invitation>,
)

@Serializable
data class InvitationsPage(val items: List<Invitation>)

@Serializable
data class CreateInvitationRequest(val email: String? = null, val username: String? = null)

interface SharingApi {
    @GET("reminders/{id}/shares")
    suspend fun listSharesAndInvitations(@Path("id") reminderId: String): SharesAndInvitationsResponse

    @POST("reminders/{id}/shares")
    suspend fun createInvitation(@Path("id") reminderId: String, @Body request: CreateInvitationRequest): Invitation

    @DELETE("reminders/{id}/shares/{shareId}")
    suspend fun revokeShare(@Path("id") reminderId: String, @Path("shareId") shareId: String)

    @DELETE("invitations/{invitationId}")
    suspend fun cancelInvitation(@Path("invitationId") invitationId: String)

    @GET("me/invitations")
    suspend fun listMyInvitations(): InvitationsPage

    @POST("invitations/{invitationId}/accept")
    suspend fun acceptInvitation(@Path("invitationId") invitationId: String): ReminderShare

    @POST("invitations/{invitationId}/reject")
    suspend fun rejectInvitation(@Path("invitationId") invitationId: String): Invitation
}

File: ios/App/Core/Network/SharingAPI.swift
import Foundation

struct ReminderShare: Codable, Identifiable {
    let id: String
    let reminderId: String
    let collaboratorUserId: String
    let status: String
    let createdAt: String
    let revokedAt: String?
}

struct Invitation: Codable, Identifiable {
    let id: String
    let reminderId: String
    let invitedEmail: String?
    let status: String
    let expiresAt: String
    let createdAt: String
}

struct SharesAndInvitationsResponse: Codable {
    let shares: [ReminderShare]
    let invitations: [Invitation]
}

private struct InvitationsPage: Decodable {
    let items: [Invitation]
}

private struct CreateInvitationRequest: Encodable {
    let email: String?
    let username: String?
}

/// AND-004/IOS-004/WEB-004 contract — mirrors android/.../SharingApi.kt and
/// web/src/features/sharing/api.ts exactly; no new endpoints invented.
enum SharingAPI {
    static func listSharesAndInvitations(reminderId: String) async throws -> SharesAndInvitationsResponse {
        let (data, response) = try await APIClient.authorizedRequest(path: "/reminders/\(reminderId)/shares", method: "GET")
        try APIClient.assertOk(response)
        return try JSONDecoder().decode(SharesAndInvitationsResponse.self, from: data)
    }

    static func createInvitation(reminderId: String, email: String?, username: String?) async throws -> Invitation {
        let body = try JSONEncoder().encode(CreateInvitationRequest(email: email, username: username))
        let (data, response) = try await APIClient.authorizedRequest(path: "/reminders/\(reminderId)/shares", method: "POST", body: body)
        try APIClient.assertOk(response)
        return try JSONDecoder().decode(Invitation.self, from: data)
    }

    static func revokeShare(reminderId: String, shareId: String) async throws {
        let (_, response) = try await APIClient.authorizedRequest(path: "/reminders/\(reminderId)/shares/\(shareId)", method: "DELETE")
        try APIClient.assertOk(response)
    }

    static func cancelInvitation(invitationId: String) async throws {
        let (_, response) = try await APIClient.authorizedRequest(path: "/invitations/\(invitationId)", method: "DELETE")
        try APIClient.assertOk(response)
    }

    static func listMyInvitations() async throws -> [Invitation] {
        let (data, response) = try await APIClient.authorizedRequest(path: "/me/invitations", method: "GET")
        try APIClient.assertOk(response)
        return try JSONDecoder().decode(InvitationsPage.self, from: data).items
    }

    static func acceptInvitation(invitationId: String) async throws {
        let (_, response) = try await APIClient.authorizedRequest(path: "/invitations/\(invitationId)/accept", method: "POST")
        try APIClient.assertOk(response)
    }

    static func rejectInvitation(invitationId: String) async throws {
        let (_, response) = try await APIClient.authorizedRequest(path: "/invitations/\(invitationId)/reject", method: "POST")
        try APIClient.assertOk(response)
    }
}
